File: graph/shortest_path/dijkstra.py
import sys
from heapq import heappush, heappop

# Fast I/O
input = sys.stdin.readline

class Dijkstra:
    """
    Dijkstra's Algorithm using Priority Queue
    Finds shortest paths from source to all other nodes
    Graph must have non-negative edge weights
    Time Complexity: O(E log V)
    """
    def __init__(self, n, adj):
        self.n = n
        self.adj = adj
        self.INF = float('inf')
        self.dist = [self.INF] * (n + 1)
        self.parent = [-1] * (n + 1)
    
    def solve(self, start_node):
        self.dist[start_node] = 0
        pq = [(0, start_node)] # (distance, node)
        
        while pq:
            d, u = heappop(pq)
            
            # Optimization: If we found a shorter path to u before, skip
            if d > self.dist[u]:
                continue
            
            for v, weight in self.adj[u]:
                if self.dist[u] + weight < self.dist[v]:
                    self.dist[v] = self.dist[u] + weight
                    self.parent[v] = u
                    heappush(pq, (self.dist[v], v))
                    
    def get_path(self, end_node):
        if self.dist[end_node] == self.INF:
            return []
        
        path = []
        curr = end_node
        while curr != -1:
            path.append(curr)
            curr = self.parent[curr]
        return path[::-1]

def solve():
    try:
        line1 = input().split()
        if not line1: return
        n, m = map(int, line1)
        
        adj = [[] for _ in range(n + 1)]
        for _ in range(m):
            # u, v, w
            u, v, w = map(int, input().split())
            adj[u].append((v, w))
            adj[v].append((u, w)) # Undirected
            
        dijkstra = Dijkstra(n, adj)
        dijkstra.solve(1)
        
        for i in range(1, n + 1):
            d = dijkstra.dist[i]
            if d == float('inf'):
                print(f"Node {i}: Unreachable")
            else:
                print(f"Node {i}: {d}")
                
    except ValueError:
        pass

if __name__ == "__main__":
    solve()

File: graph/shortest_path/dijkstra.cpp
#include <bits/stdc++.h>
using namespace std;

/*
 * Dijkstra's Algorithm
 * Shortest Path with non-negative weights
 * Time Complexity: O(E log V)
 */

const int MAXN = 200005;
const long long INF = 1e18; // Use long long for distances

struct Edge {
    int to;
    int weight;
};

vector<Edge> adj[MAXN];
long long dist[MAXN];
int parent[MAXN];

void dijkstra(int start_node, int n) {
    for (int i = 1; i <= n; i++) {
        dist[i] = INF;
        parent[i] = -1;
    }
    
    dist[start_node] = 0;
    
    // min-priority queue: (distance, node)
    priority_queue<pair<long long, int>, vector<pair<long long, int>>, greater<pair<long long, int>>> pq;
    pq.push({0, start_node});
    
    while (!pq.empty()) {
        long long d = pq.top().first;
        int u = pq.top().second;
        pq.pop();
        
        // Important: Lazy deletion check
        if (d > dist[u]) continue;
        
        for (auto& edge : adj[u]) {
            int v = edge.to;
            int w = edge.weight;
            
            if (dist[u] + w < dist[v]) {
                dist[v] = dist[u] + w;
                parent[v] = u;
                pq.push({dist[v], v});
            }
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int n, m;
    if (!(cin >> n >> m)) return 0;
    
    for (int i = 0; i < m; i++) {
        int u, v, w;
        cin >> u >> v >> w;
        adj[u].push_back({v, w});
        adj[v].push_back({u, w}); // Undirected
    }
    
    dijkstra(1, n);
    
    for (int i = 1; i <= n; i++) {
        if (dist[i] == INF) cout << "INF ";
        else cout << dist[i] << " ";
    }
    cout << "\n";
    
    return 0;
}
